libs/s25main/ogl/glSmartBitmap.cpp

```diff
@@ -255,35 +277,39 @@ void glSmartBitmap::drawPercent(DrawPoint drawPt, unsigned percent, unsigned col
             return;
     }
 
-    // nothing to draw?
-    if(!percent)
-        return;
-    RTTR_Assert(percent <= 100);
-
-    const float partDrawn = percent / 100.f;
     std::array<Point<GLfloat>, 8> vertices, curTexCoords;
     std::array<GL_RGBAColor, 8> colors;
 
+    auto drawPt = dstArea.getOrigin();
     drawPt -= origin_;
-    vertices[2] = Point<GLfloat>(drawPt) + size_;
+    vertices[2] = Point<GLfloat>(dstArea.getEndPt() - origin_); // destination bottom
 
     vertices[0].x = vertices[1].x = GLfloat(drawPt.x);
     vertices[3].x = vertices[2].x;
 
-    vertices[0].y = vertices[3].y = GLfloat(drawPt.y + size_.y * (1.f - partDrawn));
-    vertices[1].y = vertices[2].y;
+    vertices[0].y = vertices[3].y = GLfloat(drawPt.y); // destination top
+    vertices[1].y = vertices[2].y;                     // destination bottom
 
     colors[0].r = GetRed(color);
     colors[0].g = GetGreen(color);
     colors[0].b = GetBlue(color);
     colors[0].a = GetAlpha(color);
     colors[3] = colors[2] = colors[1] = colors[0];
 
+    // vertical coords
     curTexCoords[0] = texCoords[0];
     curTexCoords[1] = texCoords[1];
     curTexCoords[2] = texCoords[2];
     curTexCoords[3] = texCoords[3];
-    curTexCoords[0].y = curTexCoords[3].y = curTexCoords[1].y - (curTexCoords[1].y - curTexCoords[0].y) * partDrawn;
+    curTexCoords[0].y = curTexCoords[3].y = helpers::lerp(
+      texCoords[0].y, texCoords[1].y, helpers::inverseLerp(.0f, float(size_.y), float(srcArea.getOrigin().y)));
```

Comment:
No need for the explicit float casts (anymore). The implicit conversion is safe and the explicit casts make it harder to read.

Reply:
Still the float casts? Any reason?

Reply:
Alright in that case we can use templated lerp/inverselerp and static_assert that the type is a float value (as using an integral will make it fail due to truncation in the division)

Reply:
though the templated version doesn't change anything regarding casts - explicit `float(int)` still needs to be used or else warning will occur.

Reply:
Yes but if we need the casts anyway we can make it templated now to not have to change it in the future.